javaslang/src/main/java/javaslang/collection/Array.java

```diff
@@ -1048,11 +1048,15 @@ public int length() {
 
     @Override
     public Array<T> reverse() {
-        final Object[] arr = new Object[delegate.length];
-        for (int i = 0; i < delegate.length; i++) {
-            arr[delegate.length - 1 - i] = delegate[i];
+        if (size() <= 1) {
+            return this;
+        } else {
+            final Object[] arr = new Object[delegate.length];
+            for (int i = 0; i < delegate.length; i++) {
+                arr[delegate.length - 1 - i] = delegate[i];
+            }
```

Comment:
> The are things defined by the JVM spec that are crystal clear, e.g. heap vs. stack access.

Except that it's not :). `Escape analysis` can cheat and allocate objects on the stack, instead of the heap to unburdon the GC, see: https://docs.oracle.com/javase/8/docs/technotes/guides/vm/performance-enhancements-7.html#escapeAnalysis